### DISCUSSION TITLE
Maven Deploy fails to release nexus

### DISCUSSION THREAD

**Discussion by manikmagar** [jreleaser/jreleaser#1008] [Q&A] [answered]
Hi, I am trying to test out the new Maven Deploy with 1.3.0.

I have a simple Maven App - https://github.com/manikmagar/sample-maven-app/tree/chore/jreleaser for testing purpose. Repository contains a `jreleaser.yml` configuration file.

Deploy config - 

```
deploy:
  maven:
    nexus2:
      maven-central:
        active: ALWAYS
        url: https://oss.sonatype.org/service/local
        applyMavenCentralRules: true
        closeRepository: true
        releaseRepository: true
        stagingRepositories:
          - target/staging-deploy
```

Here is the process I followed - 
1. Stage - `mvn package deploy -Pmaven-publish -DaltDeploymentRepository=local::file:./target/staging-deploy`

2. Trigger deployment - `JRELEASER_PROJECT_VERSION=0.0.4 jreleaser deploy --config-file jreleaser.yml`

Staging of the artifact succeeds and I can see the artifact in Sonatype staging directory but it is not released. JReleaser is failing  to release the repository.

Here is a tail of the console output - 

```
[INFO]    [nexus2]  - sample-maven-app-0.0.5.pom.sha256
[INFO]    [nexus2]  - sample-maven-app-0.0.5.pom.sha512
[INFO]    [nexus2] Closing staging repository comgithubmanikmagar-1010
[INFO]    [nexus2] Releasing staging repository comgithubmanikmagar-1010
[INFO]  Writing output properties to out/jreleaser/output.properties
[ERROR] JReleaser failed after 14.930 s
feign.RetryableException: Server Error
        at org.jreleaser.sdk.nexus2.Nexus2$NexusErrorDecoder.decode(Nexus2.java:320)
        at feign.AsyncResponseHandler.handleResponse(AsyncResponseHandler.java:98)
        at feign.SynchronousMethodHandler.executeAndDecode(SynchronousMethodHandler.java:141)
        at feign.SynchronousMethodHandler.invoke(SynchronousMethodHandler.java:91)
        at feign.ReflectiveFeign$FeignInvocationHandler.invoke(ReflectiveFeign.java:100)
        at jdk.proxy2/jdk.proxy2.$Proxy17.releaseStagingRepository(Unknown Source)
        at org.jreleaser.sdk.nexus2.Nexus2.lambda$releaseStagingRepository$7(Nexus2.java:152)
        at org.jreleaser.sdk.nexus2.Nexus2.wrap(Nexus2.java:208)
        at org.jreleaser.sdk.nexus2.Nexus2.releaseStagingRepository(Nexus2.java:152)
        at org.jreleaser.sdk.nexus2.Nexus2MavenDeployer.deploy(Nexus2MavenDeployer.java:122)
......
        at org.jreleaser.cli.Main.run(Main.java:57)
        at org.jreleaser.cli.Main.main(Main.java:48)

```

Am I missing any other configuration? I am able to manually release this staged artifact from Sonatype UI. Not seeing the exact root cause of this failure. Any thoughts please? Thanks in advance!

**Comment by aalmiray** [comment]
I suppose all artifacts were checksum & signed, right? That is, there are no errors after closing the repository. 

If the repository fails to be released there should be a trace of the error in the Sonatype UI, shouldn't it?

Additionally, is there a full stacktrace at `target/jreleaser/trace.log`?

**Comment by manikmagar** [comment, marked as answer]
Thank you for the response.

Yes all signed and checksum, here is a JReleaser.yml - https://github.com/manikmagar/sample-maven-app/blob/chore/jreleaser/jreleaser.yml

from trace.log, I see it deploying asc successfully - 
```
[INFO]    [nexus2] Deploying to maven-central
[DEBUG]   [nexus2] Cannot run program "pomchecker": error=2, No such file or directory
[DEBUG]   [nexus2] cached /Users/manik/.jreleaser/caches/pomchecker/1.3.0/pomchecker-1.3.0/bin/pomchecker
[DEBUG]   [sign] signature does not exist: target/staging-deploy/com/github/manikmagar/sample-maven-app/0.0.5/sample-maven-app-0.0.5.pom.asc
[INFO]    [sign] target/staging-deploy/com/github/manikmagar/sample-maven-app/0.0.5/sample-maven-app-0.0.5.pom
[DEBUG]   [verify] target/staging-deploy/com/github/manikmagar/sample-maven-app/0.0.5/sample-maven-app-0.0.5.pom.asc
[DEBUG]   [sign] signature does not exist: target/staging-deploy/com/github/manikmagar/sample-maven-app/0.0.5/sample-maven-app-0.0.5-sources.jar.asc
[INFO]    [sign] target/staging-deploy/com/github/manikmagar/sample-maven-app/0.0.5/sample-maven-app-0.0.5-sources.jar
[DEBUG]   [verify] target/staging-deploy/com/github/manikmagar/sample-maven-app/0.0.5/sample-maven-app-0.0.5-sources.jar.asc
[DEBUG]   [sign] signature does not exist: target/staging-deploy/com/github/manikmagar/sample-maven-app/0.0.5/sample-maven-app-0.0.5.jar.asc
[INFO]    [sign] target/staging-deploy/com/github/manikmagar/sample-maven-app/0.0.5/sample-maven-app-0.0.5.jar
[DEBUG]   [verify] target/staging-deploy/com/github/manikmagar/sample-maven-app/0.0.5/sample-maven-app-0.0.5.jar.asc
[DEBUG]   [sign] signature does not exist: target/staging-deploy/com/github/manikmagar/sample-maven-app/0.0.5/sample-maven-app-0.0.5-javadoc.jar.asc
[INFO]    [sign] target/staging-deploy/com/github/manikmagar/sample-maven-app/0.0.5/sample-maven-app-0.0.5-javadoc.jar
[DEBUG]   [verify] target/staging-deploy/com/github/manikmagar/sample-maven-app/0.0.5/sample-maven-app-0.0.5-javadoc.jar.asc
[DEBUG]   [nexus2] calculating sha256 checksum for sample-maven-app-0.0.5.pom
[DEBUG]   [nexus2] calculating sha512 checksum for sample-maven-app-0.0.5.pom
[DEBUG]   [nexus2] calculating sha256 checksum for sample-maven-app-0.0.5-sources.jar
[DEBUG]   [nexus2] calculating sha512 checksum for sample-maven-app-0.0.5-sources.jar
[DEBUG]   [nexus2] calculating sha256 checksum for sample-maven-app-0.0.5.jar
[DEBUG]   [nexus2] calculating sha512 checksum for sample-maven-app-0.0.5.jar
[DEBUG]   [nexus2] calculating sha256 checksum for sample-maven-app-0.0.5-javadoc.jar
[DEBUG]   [nexus2] calculating sha512 checksum for sample-maven-app-0.0.5-javadoc.jar
[INFO]    [nexus2] Lookup staging profile for com.github.manikmagar
[DEBUG]   [nexus2] lookup staging profile for com.github.manikmagar
```

```
$ tree target/staging-deploy
target/staging-deploy
└── com
    └── github
        └── manikmagar
            └── sample-maven-app
                ├── 0.0.5
                │   ├── sample-maven-app-0.0.5-javadoc.jar
                │   ├── sample-maven-app-0.0.5-javadoc.jar.asc
                │   ├── sample-maven-app-0.0.5-javadoc.jar.md5
                │   ├── sample-maven-app-0.0.5-javadoc.jar.sha1
                │   ├── sample-maven-app-0.0.5-javadoc.jar.sha256
                │   ├── sample-maven-app-0.0.5-javadoc.jar.sha512
                │   ├── sample-maven-app-0.0.5-sources.jar
                │   ├── sample-maven-app-0.0.5-sources.jar.asc
                │   ├── sample-maven-app-0.0.5-sources.jar.md5
                │   ├── sample-maven-app-0.0.5-sources.jar.sha1
                │   ├── sample-maven-app-0.0.5-sources.jar.sha256
                │   ├── sample-maven-app-0.0.5-sources.jar.sha512
                │   ├── sample-maven-app-0.0.5.jar
                │   ├── sample-maven-app-0.0.5.jar.asc
                │   ├── sample-maven-app-0.0.5.jar.md5
                │   ├── sample-maven-app-0.0.5.jar.sha1
                │   ├── sample-maven-app-0.0.5.jar.sha256
                │   ├── sample-maven-app-0.0.5.jar.sha512
                │   ├── sample-maven-app-0.0.5.pom
                │   ├── sample-maven-app-0.0.5.pom.asc
                │   ├── sample-maven-app-0.0.5.pom.md5
                │   ├── sample-maven-app-0.0.5.pom.sha1
                │   ├── sample-maven-app-0.0.5.pom.sha256
                │   └── sample-maven-app-0.0.5.pom.sha512
                ├── maven-metadata.xml
                ├── maven-metadata.xml.md5
                └── maven-metadata.xml.sha1

5 directories, 27 files

```

There is no error on Sonatype UI. I see the closed staging directory and was able to manually release from there - 

![image](https://user-images.githubusercontent.com/877286/199218292-8268e8e6-55e2-4c1e-a77a-5c88ef565f34.png)

trace.log file (at the end) only has below stacktrace and the only error. 

```
INFO]    [nexus2] Closing staging repository comgithubmanikmagar-1010
[DEBUG]   [nexus2] closing staging repository comgithubmanikmagar-1010
[INFO]    [nexus2] Releasing staging repository comgithubmanikmagar-1010
[DEBUG]   [nexus2] releasing staging repository comgithubmanikmagar-1010
[INFO]  Writing output properties to out/jreleaser/output.properties
[ERROR] JReleaser failed after 14.930 s
feign.RetryableException: Server Error
	at org.jreleaser.sdk.nexus2.Nexus2$NexusErrorDecoder.decode(Nexus2.java:320)
	at feign.AsyncResponseHandler.handleResponse(AsyncResponseHandler.java:98)
	at feign.SynchronousMethodHandler.executeAndDecode(SynchronousMethodHandler.java:141)
	at feign.SynchronousMethodHandler.invoke(SynchronousMethodHandler.java:91)
	at feign.ReflectiveFeign$FeignInvocationHandler.invoke(ReflectiveFeign.java:100)
	at jdk.proxy2/jdk.proxy2.$Proxy17.releaseStagingRepository(Unknown Source)
	at org.jreleaser.sdk.nexus2.Nexus2.lambda$releaseStagingRepository$7(Nexus2.java:152)
	at org.jreleaser.sdk.nexus2.Nexus2.wrap(Nexus2.java:208)
	at org.jreleaser.sdk.nexus2.Nexus2.releaseStagingRepository(Nexus2.java:152)
	at org.jreleaser.sdk.nexus2.Nexus2MavenDeployer.deploy(Nexus2MavenDeployer.java:122)
	at org.jreleaser.engine.deploy.maven.ProjectMavenDeployer.deploy(ProjectMavenDeployer.java:55)
	at org.jreleaser.engine.deploy.maven.MavenDeployers.deploy(MavenDeployers.java:124)
	at org.jreleaser.engine.deploy.maven.MavenDeployers.deploy(MavenDeployers.java:111)
	at org.jreleaser.workflow.DeployWorkflowItem.doInvoke(DeployWorkflowItem.java:35)
	at org.jreleaser.workflow.AbstractWorkflowItem.lambda$invoke$0(AbstractWorkflowItem.java:43)
	at org.jreleaser.engine.hooks.CommandHookExecutor.execute(CommandHookExecutor.java:52)
	at org.jreleaser.workflow.AbstractWorkflowItem.invoke(AbstractWorkflowItem.java:43)
	at org.jreleaser.workflow.WorkflowImpl.execute(WorkflowImpl.java:102)
	at org.jreleaser.cli.Deploy.doExecute(Deploy.java:90)
	at org.jreleaser.cli.AbstractModelCommand.execute(AbstractModelCommand.java:82)
	at org.jreleaser.cli.AbstractCommand.call(AbstractCommand.java:36)
	at org.jreleaser.cli.AbstractModelCommand.call(AbstractModelCommand.java:48)
	at org.jreleaser.cli.AbstractCommand.call(AbstractCommand.java:29)
	at picocli.CommandLine.executeUserObject(CommandLine.java:1953)
	at picocli.CommandLine.access$1300(CommandLine.java:145)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2358)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2352)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2314)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2316)
	at picocli.CommandLine.execute(CommandLine.java:2078)
	at org.jreleaser.cli.Main.execute(Main.java:72)
	at org.jreleaser.cli.Main.run(Main.java:57)
	at org.jreleaser.cli.Main.main(Main.java:48)


```

**Reply by aalmiray** [reply]
If you're able to release manually it means all artifacts passed all checks and your credentials are also valid.
JReleaser has thrown an error as it encountered a 500+ response from the server, sadly the actual cause is hidden. JReleaser should probably log status, reason, and response just to be sure.

One possibility is that it takes time for a staging repository to move from open to closed state before it can actually be release, but JReleaser is not waiting at the moment. This is captured in https://github.com/jreleaser/jreleaser/issues/970

**Reply by manikmagar** [reply]
I tried releasing multiple times and error behavior is consistent. Is it possible to change the log level for Feign in cli? In case that shows any more details in request/response.

**Reply by aalmiray** [reply]
I believe `-g` is the most verbose level you may set at the moment. However, you may try passing log4j/slf4j properties using `-Dkey=value`.

**Reply by manikmagar** [reply]
`-g` did not give anything helpful for this error. 

Also, i tried with slf4j simple logger but feign did not log anything. Maybe I am not configuring it correctly. Will try again later. 
`JRELEASER_PROJECT_VERSION=0.0.5 jrel deploy -Dorg.slf4j.simpleLogger.logFile=System.out -Dorg.slf4j.simpleLogger.defaultLogLevel=debug -Dorg.slf4j.simpleLogger.log.feign=debug   --config-file jreleaser.yml`

**Reply by aalmiray** [reply]
I think Feign doesn't log that much after all. Just pushed an update (https://github.com/jreleaser/jreleaser/commit/1c6e8f063d202f5cf491295c690fe3a7a15885a9) that increases logging when a failure like this occurs. Given that you're using the CLI you may be able to test with the latest early-access (https://github.com/jreleaser/jreleaser/actions/runs/3369737829) once it's done building. Or you may build your own in this way:

0. Java 11 (and no later) is required to build
1. git clone https://github.com/jreleaser/jreleaser.git
2. cd jreleaser
3. ./gradlew :jreleaser:distZip

The distribution will be placed at `plugins/jreleaser/build/distributions`. If you're using Sdkman then you can setup a local version such as

1. unzip -d/tmp/jreleaser -qo plugins/jreleaser/build/distributions/jreleaser-1.4.0-SNAPSHOT.zip
2. sdkman i jreleaser 1.4.0-SNAPSHOT /tmp/jreleaser/jreleaser-1.4.0-SNAPSHOT.zip
3. sdkman u jreleaser 1.4.0-SNAPSHOT

**Reply by manikmagar** [reply]
I ran with the early-access zip. Thank you for a quick change to help. Now it gives some additional details but still not really the useful response from Nexus.

```
[DEBUG]   [nexus2] handling response
[INFO]    [nexus2] Closing staging repository comgithubmanikmagar-1020
[DEBUG]   [nexus2] closing staging repository comgithubmanikmagar-1020
[INFO]    [nexus2] Releasing staging repository comgithubmanikmagar-1020
[DEBUG]   [nexus2] releasing staging repository comgithubmanikmagar-1020
POST https://oss.sonatype.org/service/local/staging/profiles/355a0ce8fb3e96/promote
500 Server Error
java.io.IOException: stream is closed
	at java.base/sun.net.www.protocol.http.HttpURLConnection$HttpInputStream.ensureOpen(HttpURLConnection.java:3498)
	at java.base/sun.net.www.protocol.http.HttpURLConnection$HttpInputStream.read(HttpURLConnection.java:3523)
	at java.base/sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:284)
	at java.base/sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:326)
```

I cloned the repo locally and modified Nexus2 related code to see if issue can be resolved.

First, I added (temporarily for local only) JavaLogger for Feign builder -
```
        this.api = Feign.builder()
            .encoder(new JacksonEncoder())
                .logLevel(Logger.Level.FULL)
                .logger(new Logger.JavaLogger(Nexus2.class))
            .decoder(new ContentNegotiationDecoder())
            .requestInterceptor(new BasicAuthRequestInterceptor(username, password))
```

Then ran the CLI with `-Djava.util.logging.config.file=./log4j.properties` argument to enable `FINE` level logging for Feign. This helped to see the actual error message - 

`FINE: [NexusAPI#releaseStagingRepository] {"errors":[{"id":"*","msg":"Unhandled: Repository: comgithubmanikmagar-1026 has invalid state: open"}]}`

So, I implemented a transition verification logic that will poll until the status turns to an expected one i.e. closed. Here is a draft PR to see the change I made - https://github.com/jreleaser/jreleaser/pull/1010. 

It may not be the best polling approach but it worked 😄 and I was able to successfully make a release (see trace on PR) with this change 🎉.

Hopefully this helps.

**Reply by aalmiray** [reply]
@manikmagar thank you for the follow up and confirming my suspicions. I've been working on adding a configurable fix for retrying operations. Your PR is a good start 😄

**Reply by aalmiray** [reply]
Fixed #970. Next early-access will have the fix.

**Reply by manikmagar** [reply]
Awesome, thank you!

BTW, if later the transition fails for some reason and nexus errors need to be fetched then `GET /staging/repository/{repositoryId}/activity` could be called to extract failure reasons.

**Reply by aalmiray** [reply]
@manikmagar were you able to test with the latest early-access? Just checking if there's anything else that might be needed.

**Reply by manikmagar** [reply]
Sorry for the delay @aalmiray. Yes, tested with EA zip distribution. Release succeeded -

```
[INFO]    [nexus2]  - sample-maven-app-0.0.7.pom.sha256
[INFO]    [nexus2]  - sample-maven-app-0.0.7.pom.sha512
[INFO]    [nexus2] Closing staging repository comgithubmanikmagar-1033
[INFO]    [nexus2] Releasing staging repository comgithubmanikmagar-1033
[INFO]  Writing output properties to out/jreleaser/output.properties
[INFO]  JReleaser succeeded after 01:40 m

```

Just one observation - those repository closing and releasing tasks takes longer (30+ seconds each) and without debug it looks like it might have stuck there. Maybe the log should warn that it may take sometime to complete those tasks.

Thank you for quickly fixing it. 

QQ - Is early access also available as Github Actions? Or, that needs to wait until 1.4.0 is released?

**Reply by manikmagar** [reply]
nvm, that question of EA on GA. Found my answer here - https://github.com/jreleaser/release-action#inputs :)

**Reply by aalmiray** [reply]
I thought the same about logging a message regarding estimated time.
